services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/ReceiveSqsMessageHelper.java

```diff
@@ -0,0 +1,163 @@
+ */
+
+@SdkInternalApi
+public class ReceiveSqsMessageHelper {
+
+    private static final Logger log = Logger.loggerFor(ReceiveSqsMessageHelper.class);
+    private final String queueUrl;
+    private final SqsAsyncClient asyncClient;
+    private final Duration visibilityTimeout;
+    private final ResponseBatchConfiguration config;
+    private volatile Throwable exception;
+    private Queue<Message> messages = new ConcurrentLinkedQueue<>();
+    private volatile long visibilityDeadlineNano;
+
+    public ReceiveSqsMessageHelper(String queueUrl,
+                                   SqsAsyncClient asyncClient,
+                                   Duration visibilityTimeout,
+                                   ResponseBatchConfiguration config) {
+        this.queueUrl = queueUrl;
+        this.asyncClient = asyncClient;
+        this.visibilityTimeout = visibilityTimeout;
+        this.config = config;
+    }
+
+    public CompletableFuture<ReceiveSqsMessageHelper> asyncReceiveMessage() {
+        ReceiveMessageRequest.Builder request =
+            ReceiveMessageRequest.builder()
+                                 .queueUrl(queueUrl)
+                                 .maxNumberOfMessages(config.maxBatchItems())
+                                 .messageSystemAttributeNames(config.messageSystemAttributeNames())
+                                 .messageAttributeNames(config.receiveMessageAttributeNames());
+
+        request.visibilityTimeout(NumericUtils.saturatedCast(this.visibilityTimeout.getSeconds()));
+
+        if (config.longPollWaitTimeout() != null) {
+            request.waitTimeSeconds(NumericUtils.saturatedCast(config.longPollWaitTimeout().getSeconds()));
+        }
+        try {
+            return asyncClient.receiveMessage(request.build())
+                              .handle((response, throwable) -> {
+                                  if (throwable != null) {
+                                      setException(throwable);
+                                  } else {
+                                      messages.addAll(response.messages());
+                                  }
+                                  return this;
+                              });
+        } finally {
+            visibilityDeadlineNano = System.nanoTime() + visibilityTimeout.toNanos();
+        }
+
+    }
+
+    public boolean isEmpty() {
+        return messages == null || messages.isEmpty();
+    }
+
+    public Throwable getException() {
+        return exception;
+    }
+
+    public void setException(Throwable exception) {
+        this.exception = exception;
+    }
+
+    public Message removeMessage() {
+        if (isExpired()) {
+            clear();
+            return null;
+        }
+        return messages.poll();
+    }
+
+    private boolean isExpired() {
+        return System.nanoTime() > visibilityDeadlineNano;
+    }
+
+    public void clear() {
+        if (!isEmpty()) {
+            Optional.ofNullable(nackMessages())
```

Comment:
nit: can we just use the good ol' null check instead of wrapping it with Optional.ofNullable (extra object)?